### DISCUSSION TITLE
Is there an outline for the order of execution for each of chezmoi's features?

### DISCUSSION THREAD

**Discussion by bradenhilton** [twpayne/chezmoi#1568] [Q&A] [answered]
I've skimmed through the docs and haven't noticed anything resembling this.

An example question I had - where in the pipeline (for want of a better word) are files from `.chezmoiexternal` downloaded? During the target state generation? Can they be used in `run_before_` scripts, or are those scripts run before externals are downloaded?

Does `create_` run before `modify_`? etc.

Perhaps this could be added to the docs if it does not exist.

**Comment by twpayne** [comment, marked as answer]
The order is:

1. `run_before_` scripts in alphabetical order.
2. Targets (files, directories, scripts, symlinks, etc.) in alphabetical order of target name.
3. `run_after_` scripts in alphabetical order.

**Reply by twpayne** [reply]
Note also that the full state is read, then things are executed in order. This means that, for example, if you have a `run_before_` script that creates a `run_after_` script (shudder - you should **never** use chezmoi to modify anything in the source directory) then the `run_after_` script won't get run until then next time you run `chezmoi apply`.

**Reply by bradenhilton** [reply]
I was aware of the `run_` ordering but I suppose I erroneously assumed there would be more to it than just the target name.

Thanks for the quick response anyway.

**Reply by twpayne** [reply]
I added https://github.com/twpayne/chezmoi/pull/1584/commits/6e97b5d159fa1ebe13473592d521f73464de8a98 to include this in the docs. Thank you for the question :)

**Reply by gadams999** [reply]
Can you also do `run_once_before_`, or `run_before_once_`?

**Reply by twpayne** [reply]
The order of attributes is important. You have to use `run_once_before_`. See [the section on source state attributes in the reference manual](https://github.com/twpayne/chezmoi/blob/master/docs/REFERENCE.md#source-state-attributes) for more information.

**Comment by eugenesvk** [comment]
For SEO adding terms I initially used trying to find this: priority of commands execution
